fix: use boyer moore for efficient SVG detection

Danielku15 opened the pull request:
Fixes #397 #410
Superceeds #445

This PR ensures proper SVG size parsing for huge SVG files, trying to respect the performance concerns. 

The SVG detection is adapted to: 

* Using the Boyer Moore string search algorithm to find substrings in the provided buffer directly. This way we avoid a conversion to a utf-8 string (perf)
* The initial detection checks for the presence of `<svg` in the first 1KB which should be sufficient enough to assume SVG.  
* If no `<svg` is found, we still check if the file might be an XML file by checking for typical markers in the first 1KB. If we detect an XML we double check for `<svg` in the whole buffer.

This approach obviously brings some complexity to the code but the performance should not decrease significantly.

Danielku15 committed:
feat: boyer moore for finding byte sequences

Danielku15 committed:
fix: use boyer moore for efficient SVG detection

Danielku15 committed:
test: ensure limit is respected

AMorgaut commented:
First of all, thank you for the work, and I'm just an humble previous contributor ;-)

Looking at the given tested samples, the case you try to handle is when the svg file:

- potentially include the `<?xml version="1.0" encoding="utf-8"?>` header line
- and more impactingly include a [DTD](https://fr.wikipedia.org/wiki/Document_type_definition) which for a SVG file would start with `<!DOCTYPE svg`

So actually, if you"re looking for either `<svg` or `<!DOCTYPE svg`, you should find it within the first 100 characters, so at the very most 200 bytes

Beware that the `validate()` function is in the detection phase and then where the speed is more critical. Checks are not done in parallel ;-)

To me another point that could help to limit the impact, would be to order the tested types by their average detection speed in [types/index.ts#L23](https://github.com/image-size/image-size/blob/main/lib/types/index.ts#L23)

[detector.ts](https://github.com/image-size/image-size/blob/main/lib/detector.ts#L5) already integrates such prioritization for those detectable from their first bytes

AMorgaut commented:
Well I replied a bit fast

A file may have a svg DTD but no `<svg>` element and that's why you still look for it after...

Your choice has some logic, but in that case there should be a test case with such wrong svg file...

But then, before handling that case, let's think about what other image type could it be if it include a svg DTD. 

If there is no other valid case, then:

- no other type handler will be able to return a valid size
- it is then safe to tell the detector that you'll handle it
- and you can then look further for the `<svg>` element only once in the calculator

Danielku15 commented:
> So actually, if you"re looking for either <svg or <!DOCTYPE svg, you should find it within the first 100 characters, so at the very most 200 bytes

True, I pay much attention to how many bytes could be a potential better fit for the search. I blindly kept the 1KB. This part could be optimized further. 

> To me another point that could help to limit the impact, would be to order the tested types by their average detection speed in [types/index.ts#L23](https://github.com/image-size/image-size/blob/main/lib/types/index.ts#L23)

That would be something for the project owner to decide separately. My 2-cents on this is: It might be better to sort by likeliness of the format to be passed in. I guess JPG, PNG, SVG and WEBP are some of the most commonly used formats today and should be on top, followed by less likely used formats. 

> [detector.ts](https://github.com/image-size/image-size/blob/main/lib/detector.ts#L5) already integrates such prioritization for those detectable from their first bytes

We could maybe add SVG with `<` there. But we might have to check how a BOM impacts the loading.

> A file may have a svg DTD but no <svg> element and that's why you still look for it after...

This library only supports SVG as XML based image format. If somebody passes in a file with DTD but no svg element its safe to assume "the file is an invalid SVG" and this will throw in the `calculate` phase accordingly. 

I can add some more combinations of SVGs with and without DTDs that hit the different code paths.

tom-rm-meyer-ISST commented:
@Danielku15 and @AMorgaut, I stumbled across this issue as it has been mentioned in the docusaurus issue leading to warnings for SVGs with embedded warnings. As it's stale, I wanted to ask kindly, if there is any chance to proceed here.

As @Danielku15 indicated it would fix the docusaurus issue. I would love to get rid of the warning, if it does not affect performance (significantly). I fully understand if you and the committers / maintainers have no resources. I just wanted to give it a nudge :)

Danielku15 commented:
Seems I forgot that I wanted to follow up on this PR 😅 I'm currently travelling but might find a few mins to look into this.

Danielku15 committed:
perf(svg): detect svgs by '<' character

Danielku15 committed:
perf(svg): optimize and configurable limits

Danielku15 committed:
test(svg): cover all detection paths

Danielku15 commented:
Updated the PR with
- more specific (optimized) search limits
- allow users to adjust the search limits (by importing `SVG` and modifying the `searchLimits`)
- added more dedicated test files for the different paths.

I didn't want to blow up the code too much, hence no dedicated tests which would check whether the slow or fast path is used.

ndreno commented:
hey guys, any chance that can be reviewed? The fix looks solid and well tested

Danielku15 commented:
Ping @netroy, sorry for making noise around this. Is there anything missing to get this through?

GHF commented on lib/types/svg.ts:
```diff
@@ -86,23 +86,61 @@ function calculateByViewbox(attrs: IAttributes, viewbox: IAttributes): ISize {
+export const SVG: IImageWithSearchLimits = {
+  searchLimits: {
+    xmlStart: 10,
+    doctypeStart: 60,
```
IMO this could be a little higher (maybe 120), as this is meant to cover the case that (1) `<?xml` is missing _and_ (2) there is a `<!DOCTYPE`, in which case it's entirely reasonable to expect that there's a human readable line of comment text preceding the `<!DOCTYPE`.

`<!DOCTYPE` is also a longer needle string, which should lessen the impact of the bigger haystack, but TBH even overly large limits here will still beat the old code, considering the string decode + copy + regex match was orders of magnitude slower than your new search on the `Uint8Array`.

Danielku15 replied:
> IMO this could be a little higher (maybe 120), as this is meant to cover the case that (1) <?xml is missing and (2) there is a <!DOCTYPE, in which case it's entirely reasonable to expect that there's a human readable line of comment text preceding the <!DOCTYPE.

I haven't seen tools writing a lot of human readable comments, that's rather a project specific thing. The settings are publicly exposed for anyone who needs to tweak them for their needs in their project. `60` is a reasonable trade-off between performance and example SVGs provided in the related GH issue. 

I'm personally fine with any limit. The limits are mainly reserved due to previous PRs not making it through due to performance concerns. 

Anyhow, I fear @netroy has (temporarily?) abandoned this project and this PR will not make it through any soon. Is GH profile shows hardly any activity and the pings weren't received.

ZenaMel replied:
@Danielku15 is there anyone else besides @netroy who could put this PR through?
